providers/class.two-factor-fido-u2f.php

```diff
@@ -55,7 +147,190 @@ public function validate_authentication( $user ) {}
+		if ( ! is_numeric( $user_id ) ) {
+			return false;
+		}
+
+		$keys = get_user_meta( $user_id, self::REGISTERED_KEY_USER_META_KEY );
+		if ( $keys ) {
+			foreach ( $keys as $index => $key ) {
+				$keys[ $index ] = (object) $key;
+			}
+		}
+
+		return $keys;
+	}
+
+	/**
+	 * Update registered security key.
+	 *
+	 * Use the $prev_value parameter to differentiate between meta fields with the
+	 * same key and user ID.
+	 *
+	 * If the meta field for the user does not exist, it will be added.
+	 *
+	 * @since 0.1-dev
+	 *
+	 * @param int    $user_id  User ID.
+	 * @param object $data The data of registered security key.
+	 * @return int|bool Meta ID if the key didn't exist, true on successful update, false on failure.
+	 */
+	public static function update_security_key( $user_id, $data ) {
+		if ( ! is_numeric( $user_id ) ) {
+			return false;
+		}
+
+		if (
+			! is_object( $data )
+				|| ! property_exists( $data, 'keyHandle' ) || empty( $data->keyHandle )
+				|| ! property_exists( $data, 'publicKey' ) || empty( $data->publicKey )
+				|| ! property_exists( $data, 'certificate' ) || empty( $data->certificate )
+				|| ! property_exists( $data, 'counter' ) || ( -1 > $data->counter )
+		) {
+			return false;
+		}
+
+		$keys = get_user_meta( $user_id, self::REGISTERED_KEY_USER_META_KEY );
+		if ( $keys ) {
+			foreach ( $keys as $index => $key ) {
+				if ( $key['keyHandle'] === $data->keyHandle ) {
+					return update_user_meta( $user_id, self::REGISTERED_KEY_USER_META_KEY, (array) $data, $key );
+				}
+			}
+		}
+
+		return self::add_security_key( $user_id, $data );
+	}
+
+	/**
+	 * Remove registered security key matching criteria from a user.
+	 *
+	 * @since 0.1-dev
+	 *
+	 * @param int    $user_id   User ID.
+	 * @param string $keyHandle Optional. Key handle.
+	 * @return bool True on success, false on failure.
+	 */
+	public function delete_security_key( $user_id, $keyHandle = null ) {
+		global $wpdb;
+
+		if ( ! is_numeric( $user_id ) ) {
+			return false;
+		}
+
+		$user_id = absint( $user_id );
+		if ( ! $user_id ) {
+			return false;
+		}
+
+		$table = $wpdb->usermeta;
+
+		$keyHandle = wp_unslash( $keyHandle );
+		$keyHandle = maybe_serialize( $keyHandle );
+
+		$query = $wpdb->prepare( "SELECT umeta_id FROM $table WHERE meta_key = '%s' AND user_id = %d", self::REGISTERED_KEY_USER_META_KEY, $user_id );
+
+		if ( $keyHandle ) {
+			$query .= $wpdb->prepare( ' AND meta_value LIKE %s', '%:"' . $keyHandle . '";s:%' );
+		}
+
+		$meta_ids = $wpdb->get_col( $query );
+		if ( ! count( $meta_ids ) ) {
+			return false;
+		}
+
+		$query = "DELETE FROM $table WHERE umeta_id IN( " . implode( ',', $meta_ids ) . ' )';
+
+		$count = $wpdb->query( $query );
+
+		if ( ! $count ) {
+			return false;
+		}
+
+		wp_cache_delete( $user_id, 'user_meta' );
+
+		return true;
+	}
+
+	/**
+	 * Detect browser support for FIDO U2F.
+	 *
+	 * @since 0.1-dev
+	 */
+	public static function is_browser_support() {
```

Comment:
I'm personally uncomfortable with checking this server side, is there any way this could be an in-browser javascript check instead?  If it's not feasible, so be it, we can always revisit later.